packages/opentelemetry-propagation-utils/src/pubsub-propagation.ts

```diff
@@ -77,14 +88,30 @@ const patchArrayFunction = (
       if (!messageSpan) return callback.apply(this, callbackArgs);
 
       const res = context.with(trace.setSpan(loopContext, messageSpan), () => {
+        let result: Promise<unknown> | unknown;
         try {
-          return callback.apply(this, callbackArgs);
+          result = callback.apply(this, callbackArgs);
+          if (isPromise(result)) {
+            return result.then(
+              (x) => {
+                message[END_SPAN_FUNCTION]?.();
+                return x;
+              },
+              (err) => {
+                message[END_SPAN_FUNCTION]?.();
+                throw err;
+              }
+            );
+          }
+          return result;
```

Comment:
Here we are creating a new promise and returning it instead of the original one.
Since we are only interested in capturing the promise `resolve` / `rejected` event and do not care about modifying the result itself, I think this can be simplified to just hang another callback on the result promise:
```js
    result.then( () => message[END_SPAN_FUNCTION]?.(), () => message[END_SPAN_FUNCTION]?.())
}
return result;
```

Reply:
Good point. I've changed this.

Reply:
I think this might result in unhandled rejections now.
Calling `.then()` returns a `Promise` which is not used at all here. The catch handler throws which results in rejecting this lost promise. In the original code this chained promise was returned and the user catch handler will be called (if present).

try the following with node 14 or newer:
```js
function foo() {
  const p = Promise.reject("catch foo")
  p.then(() => {}, (e) => { throw e })
  return p
}

function bar() {
  const p = Promise.reject("catch bar")
  return p.then(() => {}, (e) => { throw e })
}

(async () => {
  try { await foo() } catch (e) {}
  try { await bar() } catch (e) {}
})()
```